core/src/main/scala/caliban/execution/Executor.scala

```diff
@@ -305,29 +305,29 @@ object Executor {
   private[caliban] def fail(error: CalibanError): UIO[GraphQLResponse[CalibanError]] =
     ZIO.succeed(GraphQLResponse(NullValue, List(error)))
 
-  private[caliban] def mergeFields(field: Field, typeName: String): List[Field] = {
-    val map      = new java.util.LinkedHashMap[String, Field]()
-    var modified = false
-
-    field.fields.foreach { field =>
-      if (field._condition.forall(_.contains(typeName))) {
-        map.compute(
-          field.aliasedName,
-          (_, f) =>
-            if (f == null) field
-            else {
-              modified = true
-              f.copy(fields = f.fields ::: field.fields)
-            }
-        )
-      } else {
-        modified = true
-      }
-    }
+  private[caliban] def mergeFields(field: Field, typeName: String): List[Field] =
+    field.fields match {
+      // Shortcut if all the fields have the same condition, which means we don't need to dedup
+      // as that's been handled in Field.apply
+      case head :: tail if tail.forall(_._condition == head._condition) =>
+        if (head._condition.forall(_.contains(typeName))) field.fields
+        else Nil
+      case Nil                                                          => Nil
+      case fields                                                       =>
+        val map = new java.util.LinkedHashMap[String, Field]()
+        fields.foreach { field =>
```

Comment:
Naked iteration I think has about a 5% perf advantage over foreach, worth doing it here? I guesss May only be noticeable with _alot_ of fields.

Reply:
Hmm good question. I'm still a bit confused on when naked iterations provide a sizeable benefit, but so far these are the case that I think it really makes a difference:

1. When it contains reference to a `var`. The Scala compiler "wraps" `var`s into a `scala.runtime.ObjectRef` because it needs to pass it to the Function1, which causes a fair bit of overhead especially if the `var` is read / modified on each iteration
2. When the list contains tuples

Although, given that this is a very hot path when using multiple fragments, I'll change it.

PS: I'll also change the `.forall` iteration above because I just realised that one causes the boolean to be boxed